AlphaWallet/TokenScriptClient/Models/FunctionOrigin.swift

```diff
@@ -257,60 +272,72 @@ struct FunctionOrigin {
         assert(functionType.isTransaction)
         let payload: Data?
         let value: BigUInt
         switch functionType {
-        case .functionCall:
+        case .functionCall, .eventFiltering:
             return nil
         case .paymentTransaction:
             payload = nil
-            guard let val = formValue(withTokenId: tokenId, attributeAndValues: attributeAndValues, server: server, account: session.account) else { return nil }
+            guard let val = formValue(withTokenId: tokenId, attributeAndValues: attributeAndValues, localRefs: localRefs, server: server, account: session.account) else { return nil }
             value = val
         case .functionTransaction:
-            guard let data = formTransactionPayload(withTokenId: tokenId, attributeAndValues: attributeAndValues, server: server, account: session.account) else { return nil }
+            guard let data = formTransactionPayload(withTokenId: tokenId, attributeAndValues: attributeAndValues, localRefs: localRefs, server: server, account: session.account) else { return nil }
             payload = data
-            value = formValue(withTokenId: tokenId, attributeAndValues: attributeAndValues, server: server, account: session.account) ?? 0
+            value = formValue(withTokenId: tokenId, attributeAndValues: attributeAndValues, localRefs: localRefs, server: server, account: session.account) ?? 0
         }
         return (payload, value)
     }
 
-    fileprivate static func extractInputs(fromDataElement dataElement: XMLElement) -> [AssetFunctionCall.Argument] {
+    fileprivate static func extractInputs(fromDataElement dataElement: XMLElement, root: XMLDocument, xmlContext: XmlContext) -> [AssetFunctionCall.Argument] {
         return XMLHandler.getInputs(fromDataElement: dataElement).compactMap {
             guard let inputType = $0.tagName.nilIfEmpty.flatMap({ SolidityType(rawValue: $0) }) else { return nil }
-            return createInput(fromInputElement: $0, withInputType: inputType)
+            return createInput(fromInputElement: $0, root: root, xmlContext: xmlContext, withInputType: inputType)
         }
     }
 
-    fileprivate static func createInput(fromInputElement inputElement: XMLElement, withInputType inputType: SolidityType) -> AssetFunctionCall.Argument? {
+    fileprivate static func createInput(fromInputElement inputElement: XMLElement, root: XMLDocument, xmlContext: XmlContext, withInputType inputType: SolidityType) -> AssetFunctionCall.Argument? {
         if let inputName = inputElement["ref"].nilIfEmpty {
             return .ref(ref: inputName, type: inputType)
+        } else if let inputName = inputElement["local-ref"].nilIfEmpty {
+            let attributes = XMLHandler.getCardAttributeTypeElements(fromRoot: root, xmlContext: xmlContext)
+            let attributeIds = attributes.compactMap { $0["id"] }
```

Comment:
@James-Sangalli The "id" introduced from the recent commit in 2020/03 might need to be changed to "name" too

Reply:
https://github.com/AlphaWallet/alpha-wallet-ios/pull/1811